Add rapid equilibrium reactions for CSTR

AntoniaBerger opened the pull request:
Fixes #20.
This PR will add the feature to set reactions into a rapid equilibrium in liquid phase for the CSTR

## Workflow:
- [x] Add a function that calculates the mass matrix with respect to conserved moieties.
- [x] Update residualImpl
 - [x] keep ordering for dynamic states
- [x] Update consitentInitalisation
  - [x] transfer from leanconsistentinitalisation to  consistent initalisation
- [x] Update Jacobian
- [x] Add AD formulation -> Ask Jan first
- [x] Update derivative Jacobian
- [x] Add interface
- [x] Add error messages
- [x] Run tests and make sure my changes don't break existing functionality.
- [ ] Add test, that ensures the functionality of the new feature.
- [ ] Update the documentation.
- [x] Make Eigen libary obligatory for CADET (already done in #471 )


More information about the algorithm are available under the [forum post about reactions in rapid equilibrium](https://forum.cadet-web.de/t/reaction-in-rapid-equilibrium/1098)

jbreue16 commented:
if you write the keyword to close the issue right in front of the corresponding issue, it will automatically be closed when the PR is merged. I.e. writing "fixes" instead of "resolves therefore" in the PR description .. maybe "resolves" also works

jbreue16 commented on src/libcadet/model/reaction/MassActionLawReaction.cpp:
```diff
@@ -26,6 +26,9 @@
 #include <string>
 #include <vector>
 
+#include <Eigen/Dense>
```
I think its time to make Eigen obligatory for CADET.
Currently, Eigen is only installed when the `ENABLE_DG` flag is set to ON (which is the default)
So when you are using Eigen in another submodule, the build will fail when `ENABLE_DG` is OFF

We can either make Eigen obligatory for CADET or you also guard the rapid-equilibrium reaction code with compile time flags checking for Eigen

jbreue16 commented:
I think its time to make Eigen obligatory for CADET.
Currently, Eigen is only installed when the `ENABLE_DG` flag is set to ON (which is the default)
So when you are using Eigen in another submodule, the build will fail when `ENABLE_DG` is OFF

We can either make Eigen obligatory for CADET or you also guard the rapid-equilibrium reaction code with compile time flags checking for Eigen.
You can see this e.g. under `/src/libcadet/CMakeLists.txt` where the DG units are only included if the flag is ON

This is only relevant for the merge into master, so for now you can also just set this as the last todo in your todo list

AntoniaBerger committed:
Restart implementation

AntoniaBerger committed:
Add conserved Moities

- to residiumImpl
- jacobian
- jacderivative

AntoniaBerger committed:
Add functionality for consisitent intialisation

AntoniaBerger committed:
Jacobian with CM is now independet from jacobian in generell case

AntoniaBerger committed:
Add interface for quasi stationary reaction in MAL

little changes

AntoniaBerger committed:
Add Consistent initalisation

AntoniaBerger committed:
Add characterisation map

AntoniaBerger committed:
Add AD functionallity

AntoniaBerger committed:
New caclulation version with eigen stare matrix multiplication instead of iteration

simple example works without consistent initalisation

fix conserved moities calculation

AntoniaBerger committed:
Running version with matrix muliplication for residium and jacobian

AntoniaBerger committed:
WIP undependent of old calculation

AntoniaBerger committed:
clean up

AntoniaBerger committed:
delete moity member from cstr and make them member of MAL

This way we are more flexible to use conserved moities in for different
units.

AntoniaBerger committed:
use index iteration

AntoniaBerger committed:
use index iteration for hess matrix

AntoniaBerger committed:
make conserved moity calculation part of mal

AntoniaBerger committed:
cm calculation only in mal module

AntoniaBerger committed:
add _dynReactiondependency to ConvDisOperator

AntoniaBerger committed:
add utility funktion

AntoniaBerger commented:
With the introduction of the new ReactionSystem.hpp and the separation of cross-phase and single-phase reactions on the master, we’re restarting the implementation of conserved moieties in a more flexible and aligned way.
